webapp/lib/main_ui.dart

```diff
@@ -45,9 +53,88 @@ class CodaUI {
     dataset.messages.forEach((message) {
       MessageViewModel messageViewModel = new MessageViewModel(message, dataset);
       messages.add(messageViewModel);
+      messageMap[message.id] = messageViewModel;
       body.append(messageViewModel.viewElement);
     });
     messageCodingTable.append(body);
+
+    messageCodingTable.onChange.listen((event) {
+      var target = event.target;
+      if (target is! InputElement && target is! SelectElement) return;
+
+      TableRowElement row = getAncestors(CodeSelector.activeCodeSelector.viewElement).firstWhere((e) => e.classes.contains('message-row'));
+      DivElement inputGroup = getAncestors(CodeSelector.activeCodeSelector.viewElement).firstWhere((e) => e.classes.contains('input-group'));
+      String messageID = row.attributes['message-id'];
+      MessageViewModel message = messageMap[messageID];
+      String schemeID = inputGroup.attributes['scheme-id'];
+
+      if (target is InputElement) { // change on checkbox element
+        message.schemeCheckChanged(dataset, schemeID, target.checked);
+      }
+      if (target is SelectElement) { // change on dropdown select element
+        CodeSelector codeSelector = message.codeSelectors.singleWhere((codeSelector) => codeSelector.scheme.id == schemeID);
+        CodeSelector.activeCodeSelector = codeSelector;
+        message.schemeCodeChanged(dataset, schemeID, codeSelector.selectedOption);
+        codeSelector.hideWarning();
+        selectNextEmptyCodeSelector(messageID, schemeID);
+      }
+    });
+
+    window.onKeyDown.listen((event) {
+      if (event.key == 'Tab') {
```

Comment:
if != tab, return.

Reply:
The if statement isn't for the whole function, only for the tab key. I'm handling other keys below. Am I missing something?

Reply:
Ah, I see, I got confused by the closure sorry